src/wiringx.c

```diff
@@ -103,6 +86,142 @@ static struct spi_t spi[2] = {
 	} while (0)
 #endif
 
+#ifndef __FreeBSD__
+/* SPI Bus Parameters */
+struct spi_t {
+	uint8_t mode;
+	uint8_t bits_per_word;
+	uint16_t delay;
+	uint32_t speed;
+	int fd;
+} spi_t;
+
+/* SPI states list implementation */
+
+struct spi_id_t {
+	uint8_t device;
+	uint8_t channel;
+};
+
+struct spi_list_t {
+	struct spi_id_t id;
+	struct spi_t data;
+};
+
+struct spi_list_t *spilist = NULL;
+int nrspilist = 0;
+
+int spi_new(struct spi_id_t id) {
+	int entry = -1;
+	void *tmp = NULL;
+	int i = 0;
+
+	// loop over existing states to detect potential duplicates
+	for(i = 0; i < nrspilist; i++) {
+		if(memcmp(&spilist[i].id, &id, sizeof(struct spi_id_t)) == 0) {
+			entry = i;
+
+			/* clear existing entry */
+			if(spilist[entry].data.fd > 0) {
+				close(spilist[entry].data.fd);
+			}
+
+			/* return index of existing entry */
+			return entry;
+		}
+	}
+
+	/* reserve space for new entry (+1) */
+	entry = nrspilist++;
+	tmp = realloc(spilist, sizeof(struct spi_list_t)*nrspilist);
+	if(tmp == NULL) {
+		fprintf(stderr, "out of memory\n");
+		exit(EXIT_FAILURE);
+	}
+	spilist = tmp;
+	spilist[entry].id = id;
+	memset(&spilist[entry].data, 0, sizeof(struct spi_t));
+
+	/* return index of new entry */
+	return entry;
+}
+
+void spi_clear() {
```

Comment:
`static`

Reply:
static? Yes and no.
Because when the test case is in a different source file, how can we call the function?
Unless when we only test the API, and not its pieces.